aggregate.h

```diff
@@ -14,81 +31,78 @@
 enum aggregate_task_type { count, sum, min, max, average /*TODO median*/ };
 
 struct aggregate_task {
-	aggregate_task_type type;
-	int from_column;
+  aggregate_task_type type;
+  int from_column;
 };
 
 template <typename T>
 struct partial_aggregate_task {
-	aggregate_task *task;
-	int iterate;
-	// mutable buffers?
-	std::vector<T> partial;
-	std::vector<T> partial1;
+  aggregate_task *task;
+  int iterate;
+  // mutable buffers?
+  std::vector<T> partial;
+  std::vector<T> partial1;
 };
 
-template <typename T>
+template<aggregate_task_type a, typename T>
 void aggregate_internal(partial_aggregate_task<T>* p_task, int row_number, int value) {
-	switch (p_task->task->type) {
-	case sum:
-		p_task->partial[row_number] += value;
-		break;
-	case count:
-		p_task->partial[row_number]++;
-		break;
-	case min:
-		if (value < p_task->partial[row_number]) {
-			p_task->partial[row_number] = value;
-		}
-		break;
-	case max:
-		if (value > p_task->partial[row_number]) {
-			p_task->partial[row_number] = value;
-		}
-		break;
-	case average:
-		p_task->partial[row_number] += value;
-		p_task->partial1[row_number]++;
-		break;
-	}
+    if constexpr(a == sum)
+        p_task->partial[row_number] += value;
+    if constexpr(a == count)
+        p_task->partial[row_number]++;
+    if constexpr(a == min)
+        if (value < p_task->partial[row_number]) {
+          p_task->partial[row_number] = value;
+        }
+    if constexpr(a == max)
+        if (value > p_task->partial[row_number]) {
+          p_task->partial[row_number] = value;
+        }
+    if constexpr(a == average) {
+        p_task->partial[row_number] += value;
+        p_task->partial1[row_number]++;
+    }
 }
 
 template <typename T, typename T2>
-void aggregate_sequential(std::shared_ptr<T2> c, group *gb, partial_aggregate_task<T>* p_task) {
-	auto cv = c->raw_values();
-	for (int j = 0; j < c->length(); j++) {
-		int row_number;
-		if (gb != NULL) {
-			row_number = gb->redirection[p_task->iterate + j];
-		} else {
-			row_number = 0;
-		}
-		aggregate_internal(p_task, row_number, cv[j]);
-	}
-	p_task->iterate += c->length();
-}
+struct aggregate_sequential {
+  template<aggregate_task_type a>
+  struct body {
+    void operator()(std::shared_ptr<T2> c, group *gb, partial_aggregate_task<T>* p_task) {
```

Comment:
honestly, I was in rush to finish it.. ideas are welcome. The obstacles are - no partial template aliasing is allowed, template cannot go into a function scope.